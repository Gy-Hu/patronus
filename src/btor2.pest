// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>

WHITESPACE = _{ " " | "\t" | "\r" }

btor2File = {SOI ~ lines ~ EOI}
lines = { ((line | WHITESPACE*) ~ NEWLINE)* }

// tokens
number = @{ ASCII_DIGIT+ }
unaryOp = @{ "not" | "inc" | "dec" | "neg" | "redand" | "redor" | "redxor" }
binaryOp = @{ "iff" | "implies" | "sgt" | "ugt" | "sgte" | "ugte" | "slt" | "ult" | "slte" | "ulte" | "and" | "nand" |
              "nor" | "or" | "xnor" | "xor" | "rol" | "ror" | "sll" | "sra" | "srl" | "add" | "mul" | "sdiv" | "udiv" |
              "smod" | "srem" | "urem" | "sub" | "saddo" | "uaddo" | "sdivo" | "udivo" | "smulo" | "umulo" | "ssubo" |
              "usubo" | "concat" | "eq" }

// each line is a declaration that is preceded by a unique ID
line = { number ~ (
declBvSort | declArraySort |
constZero | constOne | constBin | constDec | constHex | constOnes |
declState | declStateInit | declStateNext |
declBinOp |
declBad
)}

// individual commands without their ID
declBvSort = { "sort" ~ "bitvec" ~ number  }
declArraySort = { "sort" ~ "array" ~ number ~ number }
constZero = { "zero" ~ number }
constOne = { "one"  ~ number }
constBin = { "const" ~ number ~ ASCII_BIN_DIGIT* }
constDec = { "constd" ~ number ~ ASCII_DIGIT* }
constHex = { "consth" ~ number ~ ASCII_HEX_DIGIT* }
constOnes = { "ones" ~ number }
declState = { "state" ~ number  }
declStateInit = { "init" ~ number ~ number ~ number }
declStateNext = { "next" ~ number ~ number ~ number }
declBinOp = { binaryOp ~ number ~ number ~ number }
declBad = { "bad" ~ number }