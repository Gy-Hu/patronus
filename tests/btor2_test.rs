// Copyright 2023 The Regents of the University of California
// released under BSD 3-Clause License
// author: Kevin Laeufer <laeufer@berkeley.edu>

use libpatron::btor2;
use libpatron::ir::Context;
use libpatron::ir::SerializableIrNode;

const COUNT_2: &str = r#"
1 sort bitvec 3
2 zero 1
3 state 1
4 init 1 3 2
5 one 1
6 add 1 3 5
7 next 1 3 6
8 ones 1
9 sort bitvec 1
10 eq 9 3 8
11 bad 10
"#;

#[test]
fn parse_count2() {
    let mut ctx = Context::default();
    let sys = btor2::parse_str(&mut ctx, COUNT_2, Some("count2")).unwrap();
    println!("{}", sys.serialize_to_str(&ctx));
}
