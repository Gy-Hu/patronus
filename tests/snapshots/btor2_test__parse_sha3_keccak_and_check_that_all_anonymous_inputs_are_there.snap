---
source: tests/btor2_test.rs
expression: sys.serialize_to_str(&ctx)
---
sha3_keccak.w2.replace_variables
input byte_num : bv<2>
input is_last : bv<1>
input in_ready : bv<1>
input in : bv<32>
input reset : bv<1>
input clk : bv<1>
input _input_0 : bv<1>
input _input_1 : bv<1>
input _input_2 : bv<1>
input _input_3 : bv<1>
input _input_4 : bv<576>
input _input_5 : bv<576>
input _input_6 : bv<1>
input _input_7 : bv<1>
input _input_8 : bv<32>
input _input_9 : bv<7>
input _input_10 : bv<1>
input _input_11 : bv<24>
output out_ready_0 : bv<1> = i[22]
output %1788 : bv<512> = concat(f_out[1543:1536], concat(f_out[1551:1544], concat(f_out[1559:1552], concat(f_out[1567:1560], concat(f_out[1575:1568], concat(f_out[1583:1576], concat(f_out[1591:1584], concat(f_out[1599:1592], concat(f_out[1479:1472], concat(f_out[1487:1480], concat(f_out[1495:1488], concat(f_out[1503:1496], concat(f_out[1511:1504], concat(f_out[1519:1512], concat(f_out[1527:1520], concat(f_out[1535:1528], concat(f_out[1415:1408], concat(f_out[1423:1416], concat(f_out[1431:1424], concat(f_out[1439:1432], concat(f_out[1447:1440], concat(f_out[1455:1448], concat(f_out[1463:1456], concat(f_out[1471:1464], concat(f_out[1351:1344], concat(f_out[1359:1352], concat(f_out[1367:1360], concat(f_out[1375:1368], concat(f_out[1383:1376], concat(f_out[1391:1384], concat(f_out[1399:1392], concat(f_out[1407:1400], concat(f_out[1287:1280], concat(f_out[1295:1288], concat(f_out[1303:1296], concat(f_out[1311:1304], concat(f_out[1319:1312], concat(f_out[1327:1320], concat(f_out[1335:1328], concat(f_out[1343:1336], concat(f_out[1223:1216], concat(f_out[1231:1224], concat(f_out[1239:1232], concat(f_out[1247:1240], concat(f_out[1255:1248], concat(f_out[1263:1256], concat(f_out[1271:1264], concat(f_out[1279:1272], concat(f_out[1159:1152], concat(f_out[1167:1160], concat(f_out[1175:1168], concat(f_out[1183:1176], concat(f_out[1191:1184], concat(f_out[1199:1192], concat(f_out[1207:1200], concat(f_out[1215:1208], concat(f_out[1095:1088], concat(f_out[1103:1096], concat(f_out[1111:1104], concat(f_out[1119:1112], concat(f_out[1127:1120], concat(f_out[1135:1128], concat(f_out[1143:1136], f_out[1151:1144])))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))))
output _input_0 : bv<1>
node %1229 : bv<1> = ite(__synth_change_variable_19, ite(__synth_variable_135, _input_3, ite(__synth_variable_134, _input_1, ite(__synth_variable_133, state, ite(__synth_variable_132, out_ready_0, ite(__synth_variable_131, _input_0, ite(__synth_variable_130, is_last, ite(__synth_variable_129, in_ready, ite(__synth_variable_128, reset, clk)))))))), _input_2)
node %1306 : bv<1600> = ite(%1229, concat(xor(ite(__synth_change_variable_17, _input_5, _input_4), f_out[1599:1024]), f_out[1023:0]), f_out)
node %1307 : bv<64> = %1306[63:0]
node %1308 : bv<64> = %1306[383:320]
node %1309 : bv<64> = %1306[703:640]
node %1310 : bv<64> = %1306[1023:960]
node %1311 : bv<64> = %1306[1343:1280]
node %1315 : bv<64> = xor(xor(xor(xor(%1311, %1310), %1309), %1308), %1307)
node %1318 : bv<64> = concat(%1315[62:0], %1315[63])
node %1319 : bv<64> = %1306[191:128]
node %1320 : bv<64> = %1306[511:448]
node %1321 : bv<64> = %1306[831:768]
node %1322 : bv<64> = %1306[1151:1088]
node %1323 : bv<64> = %1306[1471:1408]
node %1327 : bv<64> = xor(xor(xor(xor(%1323, %1322), %1321), %1320), %1319)
node %1328 : bv<64> = %1306[1087:1024]
node %1330 : bv<64> = xor(xor(%1328, %1327), %1318)
node %1333 : bv<64> = concat(%1330[8:0], %1330[63:9])
node %1334 : bv<64> = %1306[127:64]
node %1335 : bv<64> = %1306[447:384]
node %1336 : bv<64> = %1306[767:704]
node %1337 : bv<64> = %1306[1407:1344]
node %1341 : bv<64> = xor(xor(xor(xor(%1337, %1328), %1336), %1335), %1334)
node %1344 : bv<64> = concat(%1341[62:0], %1341[63])
node %1345 : bv<64> = %1306[255:192]
node %1346 : bv<64> = %1306[575:512]
node %1347 : bv<64> = %1306[895:832]
node %1348 : bv<64> = %1306[1215:1152]
node %1349 : bv<64> = %1306[1535:1472]
node %1353 : bv<64> = xor(xor(xor(xor(%1349, %1348), %1347), %1346), %1345)
node %1355 : bv<64> = xor(xor(%1323, %1353), %1344)
node %1358 : bv<64> = concat(%1355[1:0], %1355[63:2])
node %1363 : bv<64> = concat(%1327[62:0], %1327[63])
node %1364 : bv<64> = %1306[319:256]
node %1365 : bv<64> = %1306[639:576]
node %1366 : bv<64> = %1306[959:896]
node %1367 : bv<64> = %1306[1279:1216]
node %1368 : bv<64> = %1306[1599:1536]
node %1372 : bv<64> = xor(xor(xor(xor(%1368, %1367), %1366), %1365), %1364)
node %1374 : bv<64> = xor(xor(%1345, %1372), %1363)
node %1377 : bv<64> = concat(%1374[61:0], %1374[63:62])
node %1383 : bv<64> = concat(%1353[62:0], %1353[63])
node %1385 : bv<64> = xor(xor(%1365, %1315), %1383)
node %1388 : bv<64> = concat(%1385[22:0], %1385[63:23])
node %1395 : bv<64> = concat(%1372[62:0], %1372[63])
node %1397 : bv<64> = xor(xor(%1309, %1341), %1395)
node %1400 : bv<64> = concat(%1397[24:0], %1397[63:25])
node %1412 : bv<64> = xor(xor(%1367, %1315), %1383)
node %1415 : bv<64> = concat(%1412[27:0], %1412[63:28])
node %1417 : bv<64> = xor(xor(%1311, %1341), %1395)
node %1420 : bv<64> = concat(%1417[36:0], %1417[63:37])
node %1424 : bv<64> = xor(xor(%1334, %1327), %1318)
node %1427 : bv<64> = concat(%1424[7:0], %1424[63:8])
node %1433 : bv<64> = xor(xor(%1320, %1353), %1344)
node %1436 : bv<64> = concat(%1433[48:0], %1433[63:49])
node %1442 : bv<64> = xor(xor(%1347, %1372), %1363)
node %1445 : bv<64> = concat(%1442[53:0], %1442[63:54])
node %1457 : bv<64> = xor(xor(%1322, %1353), %1344)
node %1460 : bv<64> = concat(%1457[57:0], %1457[63:58])
node %1462 : bv<64> = xor(xor(%1349, %1372), %1363)
node %1465 : bv<64> = concat(%1462[62:0], %1462[63])
node %1469 : bv<64> = xor(xor(%1364, %1315), %1383)
node %1472 : bv<64> = concat(%1469[45:0], %1469[63:46])
node %1478 : bv<64> = xor(xor(%1308, %1341), %1395)
node %1481 : bv<64> = concat(%1478[55:0], %1478[63:56])
node %1487 : bv<64> = xor(xor(%1336, %1327), %1318)
node %1490 : bv<64> = concat(%1487[38:0], %1487[63:39])
node %1502 : bv<64> = xor(xor(%1310, %1341), %1395)
node %1505 : bv<64> = concat(%1502[43:0], %1502[63:44])
node %1507 : bv<64> = xor(xor(%1337, %1327), %1318)
node %1510 : bv<64> = concat(%1507[35:0], %1507[63:36])
node %1514 : bv<64> = xor(xor(%1319, %1353), %1344)
node %1517 : bv<64> = concat(%1514[2:0], %1514[63:3])
node %1523 : bv<64> = xor(xor(%1346, %1372), %1363)
node %1526 : bv<64> = concat(%1523[18:0], %1523[63:19])
node %1532 : bv<64> = xor(xor(%1366, %1315), %1383)
node %1535 : bv<64> = concat(%1532[60:0], %1532[63:61])
node %1547 : bv<64> = xor(xor(%1348, %1372), %1363)
node %1550 : bv<64> = concat(%1547[19:0], %1547[63:20])
node %1552 : bv<64> = xor(xor(%1368, %1315), %1383)
node %1556 : bv<64> = xor(xor(%1307, %1341), %1395)
node %1559 : bv<64> = concat(%1556[49:0], %1556[63:50])
node %1565 : bv<64> = xor(xor(%1335, %1327), %1318)
node %1568 : bv<64> = concat(%1565[42:0], %1565[63:43])
node %1574 : bv<64> = xor(xor(%1321, %1353), %1344)
node %1577 : bv<64> = concat(%1574[20:0], %1574[63:21])
node %521 : bv<1> = f_permutation_.i[21]
node %519 : bv<1> = f_permutation_.i[19]
node %517 : bv<1> = f_permutation_.i[14]
node %515 : bv<1> = f_permutation_.i[13]
node %513 : bv<1> = f_permutation_.i[12]
node %511 : bv<1> = f_permutation_.i[11]
node %509 : bv<1> = f_permutation_.i[9]
node %507 : bv<1> = f_permutation_.i[6]
node %505 : bv<1> = f_permutation_.i[5]
node %503 : bv<1> = f_permutation_.i[4]
node %501 : bv<1> = f_permutation_.i[3]
node %1597 : bv<64> = xor(%1552, and(not(%1550), %1577))
node %540 : bv<1> = f_permutation_.i[18]
node %538 : bv<1> = f_permutation_.i[17]
node %536 : bv<1> = f_permutation_.i[15]
node %531 : bv<1> = f_permutation_.i[10]
node %529 : bv<1> = f_permutation_.i[7]
node %526 : bv<1> = f_permutation_.i[1]
node %527 : bv<1> = or(f_permutation_.i[0], %526)
node %528 : bv<1> = or(%527, %501)
node %556 : bv<1> = f_permutation_.i[22]
node %547 : bv<1> = f_permutation_.i[8]
node %569 : bv<1> = f_permutation_.i[20]
node %566 : bv<1> = f_permutation_.i[16]
node %573 : bv<1> = f_permutation_.i[2]
node %1248 : bv<1> = ite(__synth_change_variable_16, ite(__synth_variable_119, _input_3, ite(__synth_variable_118, _input_2, ite(__synth_variable_117, _input_1, ite(__synth_variable_116, state, ite(__synth_variable_115, out_ready_0, ite(__synth_variable_114, _input_0, ite(__synth_variable_113, is_last, ite(__synth_variable_112, in_ready, clk)))))))), reset)
node %1175 : bv<1> = ite(__synth_change_variable_7, ite(__synth_variable_62, _input_3, ite(__synth_variable_61, _input_2, ite(__synth_variable_60, _input_1, ite(__synth_variable_59, state, ite(__synth_variable_58, out_ready_0, ite(__synth_variable_57, _input_0, ite(__synth_variable_56, is_last, ite(__synth_variable_55, in_ready, clk)))))))), reset)
node %1131 : bv<1> = ite(__synth_change_variable_11, ite(__synth_variable_87, _input_3, ite(__synth_variable_86, _input_2, ite(__synth_variable_85, _input_1, ite(__synth_variable_84, state, ite(__synth_variable_83, out_ready_0, ite(__synth_variable_82, is_last, ite(__synth_variable_81, in_ready, ite(__synth_variable_80, reset, clk)))))))), _input_0)
node %1094 : bv<1> = ite(__synth_change_variable_14, ite(__synth_variable_103, _input_3, ite(__synth_variable_102, _input_1, ite(__synth_variable_101, state, ite(__synth_variable_100, out_ready_0, ite(__synth_variable_99, _input_0, ite(__synth_variable_98, is_last, ite(__synth_variable_97, in_ready, ite(__synth_variable_96, reset, clk)))))))), _input_2)
node %1095 : bv<1> = not(%1094)
node %1132 : bv<1> = not(%1131)
state i : bv<23>
  [next] ite(ite(__synth_change_variable_0, ite(__synth_variable_7, _input_3, ite(__synth_variable_6, _input_2, ite(__synth_variable_5, _input_1, ite(__synth_variable_4, state, ite(__synth_variable_3, out_ready_0, ite(__synth_variable_2, _input_0, ite(__synth_variable_1, is_last, ite(__synth_variable_0, in_ready, clk)))))))), reset), 23'x000000, concat(i[21:0], and(ite(__synth_change_variable_1, ite(__synth_variable_15, _input_3, ite(__synth_variable_14, _input_2, ite(__synth_variable_13, _input_1, ite(__synth_variable_12, out_ready_0, ite(__synth_variable_11, _input_0, ite(__synth_variable_10, is_last, ite(__synth_variable_9, in_ready, ite(__synth_variable_8, reset, clk)))))))), state), ite(__synth_change_variable_2, ite(__synth_variable_23, _input_3, ite(__synth_variable_22, _input_1, ite(__synth_variable_21, state, ite(__synth_variable_20, out_ready_0, ite(__synth_variable_19, _input_0, ite(__synth_variable_18, is_last, ite(__synth_variable_17, in_ready, ite(__synth_variable_16, reset, clk)))))))), _input_2))))
state f_out : bv<1600>
  [next] ite(%1248, 1600'x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, ite(or(f_permutation_.calc, %1229), concat(xor(%1597[63], or(or(or(or(or(or(or(or(or(or(or(or(%526, %573), %505), %507), %513), %515), %517), %536), %566), %540), %519), %569), %556)), concat(%1597[62:32], concat(xor(%1597[31], or(or(or(or(or(or(or(or(or(%573, %503), %505), %509), %531), %511), %540), %519), %521), %556)), concat(%1597[30:16], concat(xor(%1597[15], or(or(or(or(or(or(or(or(or(or(or(or(or(%527, %573), %501), %505), %507), %509), %511), %515), %517), %536), %538), %519), %569), %556)), concat(%1597[14:8], concat(xor(%1597[7], or(or(or(or(or(or(or(or(or(%528, %505), %529), %547), %511), %513), %515), %566), %519), %569)), concat(%1597[6:4], concat(xor(%1597[3], or(or(or(or(or(or(or(or(or(or(or(or(%526, %501), %507), %529), %547), %509), %531), %511), %513), %515), %538), %540), %556)), concat(%1597[2], concat(xor(%1597[1], or(or(or(or(or(or(or(or(%528, %529), %531), %511), %513), %517), %536), %538), %540)), concat(xor(%1597[0], or(or(or(or(or(or(or(or(or(or(or(%1229, %501), %503), %505), %507), %509), %511), %513), %515), %517), %519), %521)), concat(xor(%1550, and(not(%1577), %1568)), concat(xor(%1577, and(not(%1568), %1559)), concat(xor(%1568, and(not(%1559), %1552)), concat(xor(%1559, and(not(%1552), %1550)), concat(xor(%1510, and(not(%1505), %1535)), concat(xor(%1505, and(not(%1535), %1526)), concat(xor(%1535, and(not(%1526), %1517)), concat(xor(%1526, and(not(%1517), %1510)), concat(xor(%1517, and(not(%1510), %1505)), concat(xor(%1465, and(not(%1460), %1490)), concat(xor(%1460, and(not(%1490), %1481)), concat(xor(%1490, and(not(%1481), %1472)), concat(xor(%1481, and(not(%1472), %1465)), concat(xor(%1472, and(not(%1465), %1460)), concat(xor(%1420, and(not(%1415), %1445)), concat(xor(%1415, and(not(%1445), %1436)), concat(xor(%1445, and(not(%1436), %1427)), concat(xor(%1436, and(not(%1427), %1420)), concat(xor(%1427, and(not(%1420), %1415)), concat(xor(%1358, and(not(%1333), %1400)), concat(xor(%1333, and(not(%1400), %1388)), concat(xor(%1400, and(not(%1388), %1377)), concat(xor(%1388, and(not(%1377), %1358)), xor(%1377, and(not(%1358), %1333))))))))))))))))))))))))))))))))))))), f_out))
state __synth_variable_104 : bv<1>
  [next] __synth_variable_104
state __synth_variable_105 : bv<1>
  [next] __synth_variable_105
state __synth_variable_106 : bv<1>
  [next] __synth_variable_106
state __synth_variable_107 : bv<1>
  [next] __synth_variable_107
state state : bv<1>
  [next] ite(ite(__synth_change_variable_3, ite(__synth_variable_31, _input_3, ite(__synth_variable_30, _input_2, ite(__synth_variable_29, _input_1, ite(__synth_variable_28, state, ite(__synth_variable_27, out_ready_0, ite(__synth_variable_26, _input_0, ite(__synth_variable_25, is_last, ite(__synth_variable_24, in_ready, clk)))))))), reset), 1'b0, ite(ite(__synth_change_variable_4, ite(__synth_variable_39, _input_3, ite(__synth_variable_38, _input_2, ite(__synth_variable_37, _input_1, ite(__synth_variable_36, state, ite(__synth_variable_35, out_ready_0, ite(__synth_variable_34, _input_0, ite(__synth_variable_33, in_ready, ite(__synth_variable_32, reset, clk)))))))), is_last), 1'b1, state))
state __synth_variable_108 : bv<1>
  [next] __synth_variable_108
state __synth_variable_109 : bv<1>
  [next] __synth_variable_109
state __synth_variable_110 : bv<1>
  [next] __synth_variable_110
state __synth_variable_111 : bv<1>
  [next] __synth_variable_111
state __synth_change_variable_15 : bv<1>
  [next] __synth_change_variable_15
state __synth_variable_112 : bv<1>
  [next] __synth_variable_112
state __synth_variable_113 : bv<1>
  [next] __synth_variable_113
state __synth_variable_114 : bv<1>
  [next] __synth_variable_114
state __synth_variable_115 : bv<1>
  [next] __synth_variable_115
state __synth_variable_116 : bv<1>
  [next] __synth_variable_116
state __synth_variable_117 : bv<1>
  [next] __synth_variable_117
state __synth_variable_118 : bv<1>
  [next] __synth_variable_118
state __synth_variable_119 : bv<1>
  [next] __synth_variable_119
state __synth_change_variable_16 : bv<1>
  [next] __synth_change_variable_16
state __synth_change_variable_17 : bv<1>
  [next] __synth_change_variable_17
state __synth_variable_120 : bv<1>
  [next] __synth_variable_120
state __synth_variable_121 : bv<1>
  [next] __synth_variable_121
state __synth_variable_122 : bv<1>
  [next] __synth_variable_122
state __synth_variable_123 : bv<1>
  [next] __synth_variable_123
state __synth_variable_124 : bv<1>
  [next] __synth_variable_124
state __synth_variable_125 : bv<1>
  [next] __synth_variable_125
state __synth_variable_126 : bv<1>
  [next] __synth_variable_126
state __synth_variable_127 : bv<1>
  [next] __synth_variable_127
state __synth_change_variable_18 : bv<1>
  [next] __synth_change_variable_18
state __synth_variable_128 : bv<1>
  [next] __synth_variable_128
state __synth_variable_129 : bv<1>
  [next] __synth_variable_129
state __synth_variable_130 : bv<1>
  [next] __synth_variable_130
state __synth_variable_131 : bv<1>
  [next] __synth_variable_131
state __synth_variable_132 : bv<1>
  [next] __synth_variable_132
state __synth_variable_133 : bv<1>
  [next] __synth_variable_133
state __synth_variable_134 : bv<1>
  [next] __synth_variable_134
state __synth_variable_135 : bv<1>
  [next] __synth_variable_135
state __synth_change_variable_19 : bv<1>
  [next] __synth_change_variable_19
state __synth_variable_136 : bv<1>
  [next] __synth_variable_136
state __synth_variable_137 : bv<1>
  [next] __synth_variable_137
state __synth_variable_138 : bv<1>
  [next] __synth_variable_138
state __synth_variable_139 : bv<1>
  [next] __synth_variable_139
state __synth_variable_140 : bv<1>
  [next] __synth_variable_140
state __synth_variable_141 : bv<1>
  [next] __synth_variable_141
state __synth_variable_142 : bv<1>
  [next] __synth_variable_142
state __synth_variable_143 : bv<1>
  [next] __synth_variable_143
state __synth_change_variable_20 : bv<1>
  [next] __synth_change_variable_20
state f_permutation_.i : bv<23>
  [next] ite(%1248, 23'x000000, concat(f_permutation_.i[21:0], %1229))
state f_permutation_.calc : bv<1>
  [next] ite(%1248, 1'b0, or(and(f_permutation_.calc, not(%556)), %1229))
state __synth_variable_47 : bv<1>
  [next] __synth_variable_47
state __synth_variable_48 : bv<1>
  [next] __synth_variable_48
state __synth_variable_49 : bv<1>
  [next] __synth_variable_49
state __synth_variable_50 : bv<1>
  [next] __synth_variable_50
state __synth_variable_51 : bv<1>
  [next] __synth_variable_51
state __synth_variable_52 : bv<1>
  [next] __synth_variable_52
state __synth_variable_53 : bv<1>
  [next] __synth_variable_53
state __synth_variable_54 : bv<1>
  [next] __synth_variable_54
state __synth_change_variable_6 : bv<1>
  [next] __synth_change_variable_6
state __synth_variable_55 : bv<1>
  [next] __synth_variable_55
state __synth_variable_56 : bv<1>
  [next] __synth_variable_56
state __synth_variable_57 : bv<1>
  [next] __synth_variable_57
state __synth_variable_58 : bv<1>
  [next] __synth_variable_58
state __synth_variable_59 : bv<1>
  [next] __synth_variable_59
state __synth_variable_60 : bv<1>
  [next] __synth_variable_60
state __synth_variable_61 : bv<1>
  [next] __synth_variable_61
state __synth_variable_62 : bv<1>
  [next] __synth_variable_62
state __synth_change_variable_7 : bv<1>
  [next] __synth_change_variable_7
state __synth_variable_63 : bv<1>
  [next] __synth_variable_63
state __synth_change_variable_8 : bv<1>
  [next] __synth_change_variable_8
state __synth_variable_64 : bv<1>
  [next] __synth_variable_64
state __synth_variable_65 : bv<1>
  [next] __synth_variable_65
state __synth_variable_66 : bv<1>
  [next] __synth_variable_66
state __synth_variable_67 : bv<1>
  [next] __synth_variable_67
state __synth_variable_68 : bv<1>
  [next] __synth_variable_68
state __synth_variable_69 : bv<1>
  [next] __synth_variable_69
state __synth_variable_70 : bv<1>
  [next] __synth_variable_70
state __synth_variable_71 : bv<1>
  [next] __synth_variable_71
state __synth_change_variable_9 : bv<1>
  [next] __synth_change_variable_9
state __synth_variable_72 : bv<1>
  [next] __synth_variable_72
state __synth_variable_73 : bv<1>
  [next] __synth_variable_73
state __synth_variable_74 : bv<1>
  [next] __synth_variable_74
state __synth_variable_75 : bv<1>
  [next] __synth_variable_75
state __synth_variable_76 : bv<1>
  [next] __synth_variable_76
state __synth_variable_77 : bv<1>
  [next] __synth_variable_77
state __synth_variable_78 : bv<1>
  [next] __synth_variable_78
state __synth_variable_79 : bv<1>
  [next] __synth_variable_79
state __synth_change_variable_10 : bv<1>
  [next] __synth_change_variable_10
state __synth_variable_80 : bv<1>
  [next] __synth_variable_80
state __synth_variable_81 : bv<1>
  [next] __synth_variable_81
state __synth_variable_82 : bv<1>
  [next] __synth_variable_82
state __synth_variable_83 : bv<1>
  [next] __synth_variable_83
state __synth_variable_84 : bv<1>
  [next] __synth_variable_84
state __synth_variable_85 : bv<1>
  [next] __synth_variable_85
state __synth_variable_86 : bv<1>
  [next] __synth_variable_86
state __synth_variable_87 : bv<1>
  [next] __synth_variable_87
state __synth_change_variable_11 : bv<1>
  [next] __synth_change_variable_11
state __synth_change_variable_12 : bv<1>
  [next] __synth_change_variable_12
state __synth_variable_96 : bv<1>
  [next] __synth_variable_96
state __synth_variable_97 : bv<1>
  [next] __synth_variable_97
state __synth_variable_98 : bv<1>
  [next] __synth_variable_98
state __synth_variable_99 : bv<1>
  [next] __synth_variable_99
state __synth_variable_100 : bv<1>
  [next] __synth_variable_100
state __synth_variable_101 : bv<1>
  [next] __synth_variable_101
state __synth_variable_102 : bv<1>
  [next] __synth_variable_102
state __synth_variable_103 : bv<1>
  [next] __synth_variable_103
state __synth_change_variable_14 : bv<1>
  [next] __synth_change_variable_14
state padder_.state : bv<1>
  [next] ite(%1175, 1'b0, ite(ite(__synth_change_variable_10, ite(__synth_variable_79, _input_3, ite(__synth_variable_78, _input_2, ite(__synth_variable_77, _input_1, ite(__synth_variable_76, state, ite(__synth_variable_75, out_ready_0, ite(__synth_variable_74, _input_0, ite(__synth_variable_73, in_ready, ite(__synth_variable_72, reset, clk)))))))), is_last), 1'b1, padder_.state))
state padder_.done : bv<1>
  [next] ite(%1175, 1'b0, ite(and(padder_.state, %1131), 1'b1, padder_.done))
state padder_.i : bv<18>
  [next] ite(%1175, 18'x00000, ite(or(%1094, and(or(and(and(not(padder_.state), ite(__synth_change_variable_9, ite(__synth_variable_71, _input_3, ite(__synth_variable_70, _input_2, ite(__synth_variable_69, _input_1, ite(__synth_variable_68, state, ite(__synth_variable_67, out_ready_0, ite(__synth_variable_66, _input_0, ite(__synth_variable_65, is_last, ite(__synth_variable_64, reset, clk)))))))), in_ready)), %1132), and(padder_.state, %1132)), not(padder_.done))), and(concat(padder_.i[16:0], 1'b1), concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, concat(%1095, %1095)))))))))))))))))), padder_.i))
state __synth_variable_95 : bv<1>
  [next] __synth_variable_95
state __synth_variable_94 : bv<1>
  [next] __synth_variable_94
state __synth_variable_93 : bv<1>
  [next] __synth_variable_93
state __synth_variable_92 : bv<1>
  [next] __synth_variable_92
state __synth_variable_91 : bv<1>
  [next] __synth_variable_91
state __synth_variable_90 : bv<1>
  [next] __synth_variable_90
state __synth_variable_89 : bv<1>
  [next] __synth_variable_89
state __synth_variable_88 : bv<1>
  [next] __synth_variable_88
state __synth_variable_46 : bv<1>
  [next] __synth_variable_46
state __synth_variable_45 : bv<1>
  [next] __synth_variable_45
state __synth_variable_44 : bv<1>
  [next] __synth_variable_44
state __synth_variable_43 : bv<1>
  [next] __synth_variable_43
state __synth_variable_42 : bv<1>
  [next] __synth_variable_42
state __synth_variable_41 : bv<1>
  [next] __synth_variable_41
state __synth_variable_40 : bv<1>
  [next] __synth_variable_40
state __synth_variable_39 : bv<1>
  [next] __synth_variable_39
state __synth_variable_38 : bv<1>
  [next] __synth_variable_38
state __synth_variable_37 : bv<1>
  [next] __synth_variable_37
state __synth_variable_36 : bv<1>
  [next] __synth_variable_36
state __synth_variable_35 : bv<1>
  [next] __synth_variable_35
state __synth_variable_34 : bv<1>
  [next] __synth_variable_34
state __synth_variable_33 : bv<1>
  [next] __synth_variable_33
state __synth_variable_32 : bv<1>
  [next] __synth_variable_32
state __synth_variable_31 : bv<1>
  [next] __synth_variable_31
state __synth_variable_30 : bv<1>
  [next] __synth_variable_30
state __synth_variable_29 : bv<1>
  [next] __synth_variable_29
state __synth_variable_28 : bv<1>
  [next] __synth_variable_28
state __synth_variable_27 : bv<1>
  [next] __synth_variable_27
state __synth_variable_26 : bv<1>
  [next] __synth_variable_26
state __synth_variable_25 : bv<1>
  [next] __synth_variable_25
state __synth_variable_24 : bv<1>
  [next] __synth_variable_24
state __synth_variable_23 : bv<1>
  [next] __synth_variable_23
state __synth_variable_22 : bv<1>
  [next] __synth_variable_22
state __synth_variable_21 : bv<1>
  [next] __synth_variable_21
state __synth_variable_20 : bv<1>
  [next] __synth_variable_20
state __synth_variable_19 : bv<1>
  [next] __synth_variable_19
state __synth_variable_18 : bv<1>
  [next] __synth_variable_18
state __synth_variable_17 : bv<1>
  [next] __synth_variable_17
state __synth_variable_16 : bv<1>
  [next] __synth_variable_16
state __synth_variable_15 : bv<1>
  [next] __synth_variable_15
state __synth_variable_14 : bv<1>
  [next] __synth_variable_14
state __synth_variable_13 : bv<1>
  [next] __synth_variable_13
state __synth_variable_12 : bv<1>
  [next] __synth_variable_12
state __synth_variable_11 : bv<1>
  [next] __synth_variable_11
state __synth_variable_10 : bv<1>
  [next] __synth_variable_10
state __synth_variable_9 : bv<1>
  [next] __synth_variable_9
state __synth_variable_8 : bv<1>
  [next] __synth_variable_8
state __synth_variable_7 : bv<1>
  [next] __synth_variable_7
state __synth_variable_6 : bv<1>
  [next] __synth_variable_6
state __synth_variable_5 : bv<1>
  [next] __synth_variable_5
state __synth_variable_4 : bv<1>
  [next] __synth_variable_4
state __synth_variable_3 : bv<1>
  [next] __synth_variable_3
state __synth_variable_2 : bv<1>
  [next] __synth_variable_2
state __synth_variable_1 : bv<1>
  [next] __synth_variable_1
state __synth_variable_0 : bv<1>
  [next] __synth_variable_0
state __synth_change_variable_13 : bv<1>
  [next] __synth_change_variable_13
state __synth_change_variable_5 : bv<1>
  [next] __synth_change_variable_5
state __synth_change_variable_4 : bv<1>
  [next] __synth_change_variable_4
state __synth_change_variable_3 : bv<1>
  [next] __synth_change_variable_3
state __synth_change_variable_2 : bv<1>
  [next] __synth_change_variable_2
state __synth_change_variable_1 : bv<1>
  [next] __synth_change_variable_1
state __synth_change_variable_0 : bv<1>
  [next] __synth_change_variable_0

